In [7]:
import os
import glob

base = os.path.expanduser("~/IDX-Exchange-Internship-Data_Science-Summer2026/california")

files = glob.glob(os.path.join(base, "CRMLSSold*.csv"))

print("Base folder:", base)
print("Folder exists:", os.path.exists(base))
print("Files found:", len(files))
print(files[:5])

Base folder: /Users/nguyenanh/IDX-Exchange-Internship-Data_Science-Summer2026/california
Folder exists: True
Files found: 30
['/Users/nguyenanh/IDX-Exchange-Internship-Data_Science-Summer2026/california/CRMLSSold202404_filled.csv', '/Users/nguyenanh/IDX-Exchange-Internship-Data_Science-Summer2026/california/CRMLSSold202409.csv', '/Users/nguyenanh/IDX-Exchange-Internship-Data_Science-Summer2026/california/CRMLSSold202408.csv', '/Users/nguyenanh/IDX-Exchange-Internship-Data_Science-Summer2026/california/CRMLSSold202501_filled.csv', '/Users/nguyenanh/IDX-Exchange-Internship-Data_Science-Summer2026/california/CRMLSSold20220101_20231231_filled.csv']


In [8]:
# Load all files
dfs = []
for f in sorted(files):
    temp = pd.read_csv(f, low_memory=False)
    temp["_source_file"] = os.path.basename(f)  # track which file each row came from
    print(f"{os.path.basename(f):50s} → {len(temp):>8,} rows, {temp.shape[1]} cols")
    dfs.append(temp)

df_raw = pd.concat(dfs, ignore_index=True)
print(f"\nTotal rows: {len(df_raw):,}")
print(f"Total columns: {df_raw.shape[1]}")

CRMLSSold20220101_20231231_filled.csv              →  157,828 rows, 81 cols
CRMLSSold202401_filled.csv                         →   17,958 rows, 81 cols
CRMLSSold202402_filled.csv                         →   19,925 rows, 81 cols
CRMLSSold202403_filled.csv                         →   23,276 rows, 81 cols
CRMLSSold202404_filled.csv                         →   24,640 rows, 81 cols
CRMLSSold202405_filled.csv                         →   26,487 rows, 81 cols
CRMLSSold202406_filled.csv                         →   24,328 rows, 81 cols
CRMLSSold202407_filled.csv                         →   26,240 rows, 81 cols
CRMLSSold202408.csv                                →   24,558 rows, 79 cols
CRMLSSold202409.csv                                →   21,267 rows, 79 cols
CRMLSSold202410.csv                                →   23,274 rows, 79 cols
CRMLSSold202411.csv                                →   20,279 rows, 79 cols
CRMLSSold202412.csv                                →   20,241 rows, 79 cols
CRMLSSold202

In [9]:
df_raw["_source_file"].head()

0    CRMLSSold20220101_20231231_filled.csv
1    CRMLSSold20220101_20231231_filled.csv
2    CRMLSSold20220101_20231231_filled.csv
3    CRMLSSold20220101_20231231_filled.csv
4    CRMLSSold20220101_20231231_filled.csv
Name: _source_file, dtype: object

In [10]:
df_raw.head()

,Flooring,ViewYN,WaterfrontYN,BasementYN,PoolPrivateYN,OriginalListPrice,ListingKey,ListAgentEmail,CloseDate,ClosePrice,...,HighSchoolDistrict,PostalCode,AssociationFee,LotSizeSquareFeet,MiddleOrJuniorSchoolDistrict,latfilled,lonfilled,_source_file,BuyerAgentAOR,ListAgentAOR
0,"Carpet,Wood",True,NaN,NaN,True,98000.0,556366533,michellefsellsoc@gmail.com,2022-02-25,95000.0,...,ABC Unified,92708,0.0,NaN,NaN,False,False,CRMLSSold20220101_20231231_filled.csv,NaN,NaN
1,NaN,False,NaN,NaN,False,1200.0,556366530,dineshcalre@gmail.com,2022-02-19,1200.0,...,Apple Valley Unified,92308,0.0,43000.0,NaN,False,False,CRMLSSold20220101_20231231_filled.csv,NaN,NaN
2,NaN,True,NaN,NaN,False,1100000.0,556366044,cindydavishomes@gmail.com,2022-04-15,1100000.0,...,Solana Beach,92075,370.0,NaN,NaN,False,False,CRMLSSold20220101_20231231_filled.csv,NaN,NaN
3,NaN,True,NaN,NaN,False,2499999.0,556365765,bryanmeathe@gmail.com,2022-01-04,2499999.0,...,Carlsbad Unified,92008,140.0,13376.0,NaN,False,False,CRMLSSold20220101_20231231_filled.csv,NaN,NaN
4,"Carpet,Tile",NaN,NaN,NaN,NaN,598888.0,556365290,steven@westsideres.com,2022-01-12,640000.0,...,Other,95111,300.0,2738.0,NaN,False,False,CRMLSSold20220101_20231231_filled.csv,NaN,NaN


In [11]:
df_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 794271 entries, 0 to 794270
Data columns (total 83 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   Flooring                      465886 non-null  object 
 1   ViewYN                        707277 non-null  object 
 2   WaterfrontYN                  434 non-null     object 
 3   BasementYN                    13172 non-null   object 
 4   PoolPrivateYN                 690017 non-null  object 
 5   OriginalListPrice             791935 non-null  float64
 6   ListingKey                    794271 non-null  int64  
 7   ListAgentEmail                792478 non-null  object 
 8   CloseDate                     794271 non-null  object 
 9   ClosePrice                    794263 non-null  float64
 10  ListAgentFirstName            789394 non-null  object 
 11  ListAgentLastName             794180 non-null  object 
 12  Latitude                      793455 non-nul

In [12]:
df_raw[["ClosePrice", "OriginalListPrice", "LivingArea", "BedroomsTotal", 
        "BathroomsTotalInteger", "DaysOnMarket", "YearBuilt"]].describe()

,ClosePrice,OriginalListPrice,LivingArea,BedroomsTotal,BathroomsTotalInteger,DaysOnMarket,YearBuilt
count,7.942630e+05,7.919350e+05,7.366580e+05,739759.000000,755771.00000,794270.000000,759347.000000
mean,8.675752e+05,9.027827e+05,5.539220e+03,3.077731,2.44057,42.427458,1977.914961
std,5.044244e+06,5.827031e+06,1.834682e+06,1.374562,1.38339,75.919275,26.998957
min,0.000000e+00,0.000000e+00,0.000000e+00,0.000000,0.00000,-288.000000,1776.000000
25%,5.127500e+04,6.499900e+04,1.183000e+03,2.000000,2.00000,8.000000,1959.000000
50%,6.250000e+05,6.330000e+05,1.579000e+03,3.000000,2.00000,20.000000,1979.000000
75%,1.060000e+06,1.074900e+06,2.149000e+03,4.000000,3.00000,50.000000,1999.000000
max,9.895000e+08,1.390000e+09,9.090991e+08,180.000000,175.00000,12430.000000,2026.000000


In [17]:
df_raw.isna().sum().sort_values(ascending=False).head(20)

CoveredSpaces                   794271
AboveGradeFinishedArea          794271
FireplacesTotal                 794271
MiddleOrJuniorSchoolDistrict    794271
ElementarySchoolDistrict        794271
WaterfrontYN                    793837
TaxYear                         793773
BusinessType                    792013
BelowGradeFinishedArea          790928
TaxAnnualAmount                 790912
BasementYN                      781099
BuilderName                     764093
LotSizeDimensions               752083
CoBuyerAgentFirstName           730015
ElementarySchool                704646
MiddleOrJuniorSchool            703918
BuildingAreaTotal               693048
HighSchool                      677564
CoListAgentFirstName            631672
CoListAgentLastName             631351
dtype: int64

In [18]:
df_raw["ClosePrice"].isna().sum()

np.int64(8)

In [19]:
df_raw["ListingKey"].duplicated().sum()

np.int64(646)

In [11]:
# Check column consistency across files
col_sets = {os.path.basename(f): set(d.columns) for f, d in zip(sorted(files), dfs)}
all_cols = set.union(*col_sets.values())

print(f"Total unique columns across all files: {len(all_cols)}\n")

any_mismatch = False
for fname, cols in col_sets.items():
    missing = all_cols - cols
    extra   = cols - all_cols  # won't happen but good to check
    if missing:
        any_mismatch = True
        print(f"⚠️  {fname} MISSING {len(missing)} cols: {missing}")

if not any_mismatch:
    print("✅ All files have identical columns — clean concat!")

Total unique columns across all files: 83

⚠️  CRMLSSold20220101_20231231_filled.csv MISSING 2 cols: {'ListAgentAOR', 'BuyerAgentAOR'}
⚠️  CRMLSSold202401_filled.csv MISSING 2 cols: {'ListAgentAOR', 'BuyerAgentAOR'}
⚠️  CRMLSSold202402_filled.csv MISSING 2 cols: {'ListAgentAOR', 'BuyerAgentAOR'}
⚠️  CRMLSSold202403_filled.csv MISSING 2 cols: {'ListAgentAOR', 'BuyerAgentAOR'}
⚠️  CRMLSSold202404_filled.csv MISSING 2 cols: {'ListAgentAOR', 'BuyerAgentAOR'}
⚠️  CRMLSSold202405_filled.csv MISSING 2 cols: {'BuyerAgencyCompensation', 'BuyerAgencyCompensationType'}
⚠️  CRMLSSold202406_filled.csv MISSING 2 cols: {'BuyerAgencyCompensation', 'BuyerAgencyCompensationType'}
⚠️  CRMLSSold202407_filled.csv MISSING 2 cols: {'BuyerAgencyCompensation', 'BuyerAgencyCompensationType'}
⚠️  CRMLSSold202408.csv MISSING 4 cols: {'lonfilled', 'latfilled', 'BuyerAgencyCompensation', 'BuyerAgencyCompensationType'}
⚠️  CRMLSSold202409.csv MISSING 4 cols: {'lonfilled', 'latfilled', 'BuyerAgencyCompensation', 'Buy

In [12]:
# Week 1 sanity check — just confirm data loaded correctly
print(f"Rows:    {len(df_raw):,}")
print(f"Columns: {df_raw.shape[1]}")
print(f"\nColumn names:")
print(list(df_raw.columns))

Rows:    794,271
Columns: 83

Column names:
['Flooring', 'ViewYN', 'WaterfrontYN', 'BasementYN', 'PoolPrivateYN', 'OriginalListPrice', 'ListingKey', 'ListAgentEmail', 'CloseDate', 'ClosePrice', 'ListAgentFirstName', 'ListAgentLastName', 'Latitude', 'Longitude', 'UnparsedAddress', 'PropertyType', 'LivingArea', 'ListPrice', 'DaysOnMarket', 'ListOfficeName', 'BuyerOfficeName', 'CoListOfficeName', 'ListAgentFullName', 'CoListAgentFirstName', 'CoListAgentLastName', 'BuyerAgentMlsId', 'BuyerAgentFirstName', 'BuyerAgentLastName', 'FireplacesTotal', 'AssociationFeeFrequency', 'AboveGradeFinishedArea', 'ListingKeyNumeric', 'MLSAreaMajor', 'TaxAnnualAmount', 'CountyOrParish', 'MlsStatus', 'ElementarySchool', 'AttachedGarageYN', 'ParkingTotal', 'BuilderName', 'PropertySubType', 'LotSizeAcres', 'SubdivisionName', 'BuyerOfficeAOR', 'YearBuilt', 'BuyerAgencyCompensationType', 'StreetNumberNumeric', 'ListingId', 'BathroomsTotalInteger', 'City', 'BuyerAgencyCompensation', 'TaxYear', 'BuildingAreaTotal

In [20]:
for i, col in enumerate(df_raw.columns, start=1):
    print(i, col)

1 Flooring
2 ViewYN
3 WaterfrontYN
4 BasementYN
5 PoolPrivateYN
6 OriginalListPrice
7 ListingKey
8 ListAgentEmail
9 CloseDate
10 ClosePrice
11 ListAgentFirstName
12 ListAgentLastName
13 Latitude
14 Longitude
15 UnparsedAddress
16 PropertyType
17 LivingArea
18 ListPrice
19 DaysOnMarket
20 ListOfficeName
21 BuyerOfficeName
22 CoListOfficeName
23 ListAgentFullName
24 CoListAgentFirstName
25 CoListAgentLastName
26 BuyerAgentMlsId
27 BuyerAgentFirstName
28 BuyerAgentLastName
29 FireplacesTotal
30 AssociationFeeFrequency
31 AboveGradeFinishedArea
32 ListingKeyNumeric
33 MLSAreaMajor
34 TaxAnnualAmount
35 CountyOrParish
36 MlsStatus
37 ElementarySchool
38 AttachedGarageYN
39 ParkingTotal
40 BuilderName
41 PropertySubType
42 LotSizeAcres
43 SubdivisionName
44 BuyerOfficeAOR
45 YearBuilt
46 BuyerAgencyCompensationType
47 StreetNumberNumeric
48 ListingId
49 BathroomsTotalInteger
50 City
51 BuyerAgencyCompensation
52 TaxYear
53 BuildingAreaTotal
54 BedroomsTotal
55 ContractStatusChangeDate
56 Ele

In [27]:

# CRMLS key column dictionary
column_meanings = {
    # --- Price & sale outcome ---
    "ClosePrice":            "Final sale price (main target variable)",
    "ListPrice":             "Most recent list price",
    "OriginalListPrice":     "Price the property was first listed at",
    "DaysOnMarket":          "Days active before going under contract",
    # --- Property characteristics ---
    "PropertyType":          "Top-level type (Residential, Commercial, Land…)",
    "PropertySubType":       "Detailed type (Single Family Residence, Condominium…)",
    "LivingArea":            "Finished living area (sq ft)",
    "LotSizeSquareFeet":     "Lot size in square feet",
    "BedroomsTotal":         "Total bedrooms",
    "BathroomsTotalInteger": "Total bathrooms (whole number)",
    "YearBuilt":             "Year built",
    "Stories":               "Number of stories",
    "GarageSpaces":          "Garage parking spaces",
    "PoolPrivateYN":         "Has a private pool (Yes/No)",
    "ViewYN":                "Has a view (Yes/No)",
    "FireplacesTotal":       "Number of fireplaces",
    "NewConstructionYN":     "New construction (Yes/No)",
    # --- Location ---
    "City":                  "City",
    "PostalCode":            "ZIP code",
    "CountyOrParish":        "County",
    "Latitude":              "Property latitude",
    "Longitude":             "Property longitude",
    "UnparsedAddress":       "Full street address",
    "MLSAreaMajor":          "Major MLS geographic area",
    # --- Costs / taxes ---
    "TaxAnnualAmount":       "Annual property tax",
    "AssociationFee":        "HOA fee amount",
    "AssociationFeeFrequency": "HOA fee billing frequency (Monthly, Annually…)",
    # --- Timing ---
    "CloseDate":             "Date the sale closed",
    "ListingContractDate":   "Date listed for sale",
    "PurchaseContractDate":  "Date the offer was accepted (went into escrow)",
    # --- Identifiers ---
    "ListingId":             "Human-readable MLS listing number",
    "ListingKey":            "Unique system ID for the listing record",
}

col_dict = pd.DataFrame(column_meanings.items(), columns=["Column", "Meaning"])



In [28]:
col_dict["In data"]   = col_dict["Column"].isin(df_raw.columns)
col_dict["Non-null %"] = col_dict["Column"].map(
    lambda c: round(df_raw[c].notna().mean() * 100, 1) if c in df_raw.columns else None
)
col_dict


,Column,Meaning,In data,Non-null %
0,ClosePrice,Final sale price (main target variable),True,100.0
1,ListPrice,Most recent list price,True,99.8
2,OriginalListPrice,Price the property was first listed at,True,99.7
3,DaysOnMarket,Days active before going under contract,True,100.0
4,PropertyType,"Top-level type (Residential, Commercial, Land…)",True,100.0
5,PropertySubType,"Detailed type (Single Family Residence, Condom...",True,92.1
6,LivingArea,Finished living area (sq ft),True,92.7
7,LotSizeSquareFeet,Lot size in square feet,True,90.7
8,BedroomsTotal,Total bedrooms,True,93.1
9,BathroomsTotalInteger,Total bathrooms (whole number),True,95.2


### Week 1 Dataset Sanity Check

I successfully loaded the California CRMLS Sold dataset into Python. The combined dataset contains 794,271 rows and 83 columns, which confirms that the files were accessed and combined correctly.

The dataset includes important columns for price prediction, such as `ClosePrice`, `OriginalListPrice`, `LivingArea`, `BedroomsTotal`, `BathroomsTotalInteger`, `DaysOnMarket`, `YearBuilt`, `PropertyType`, `Latitude`, `Longitude`, `CloseDate`, and `ListingKey`.

The target variable for the project is likely `ClosePrice`, which represents the final sale price of the property. This column is mostly complete, with only 8 missing values out of 794,271 rows.

I also found some data quality issues that will need to be handled later. Some numeric columns contain unrealistic or invalid values, such as `ClosePrice` equal to 0, `OriginalListPrice` equal to 0, `LivingArea` equal to 0, negative `DaysOnMarket`, and unusually high bedroom or bathroom counts. These values should be cleaned before modeling.

There are also 646 duplicate `ListingKey` values, which means some listings appear more than once. These duplicates should be removed during the cleaning stage.

Some columns have many missing values, including `CoveredSpaces`, `AboveGradeFinishedArea`, `FireplacesTotal`, school district columns, `WaterfrontYN`, `TaxYear`, `BusinessType`, and `BasementYN`. These columns may not be useful for the first version of the model unless they are cleaned carefully.

Overall, the dataset was loaded successfully and contains the main features needed to begin exploratory data analysis and cleaning.